src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt

```diff
@@ -79,8 +89,27 @@ class AntiSpam : Module() {
                 .filter { entry: Map.Entry<String, Long> -> entry.value < System.currentTimeMillis() - 10 * 60 * 1000 } // 10 is delay in minutes
                 .collect(Collectors.toList())
                 .forEach(Consumer { entry: Map.Entry<String, Long> -> messageHistory!!.remove(entry.key) })
-        if (isSpam(event.message.unformattedText)) {
-            event.isCanceled = true
+
+        val pattern = isSpam(event.message.unformattedText)
+        if (pattern != null) { // null means no pattern found
+            if (mode.value == Mode.HIDE) {
+                event.isCanceled = true
+            } else if (mode.value == Mode.REPLACE) {
+                event.message = TextComponentString(sanitize(event.message.formattedText, pattern, (replaceMode.value as ReplaceMode).redaction))
+            }
+        }
+
+        if (duplicates.value) {
+            if (checkDupes(event.message.unformattedText)) {
+                event.isCanceled = true
+            }
+        }
```

Comment:
I'd favor duplicate checking over censoring, can you move this before the pattern check?

Reply:
sounds good!